src/Service/ExecuteSourceDeductionForItems.php

```diff
@@ -111,13 +119,16 @@ public function executeSourceDeductionForItems(OrderItem $orderItem, array $item
         ]);
 
         $salesEvent = $this->salesEventFactory->create([
-            'type' => $type,
+            'type' => SalesEventInterface::EVENT_ORDER_CANCELED,
             'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
-            'objectId' => (string)$order->getId()
+            'objectId' => (string)$orderItem->getOrderId()
         ]);
 
+        $itemsIds = [];
+
         /** @var OrderItem $item */
         foreach ($itemsToCancel as $item) {
+            $itemsIds[] = $this->productRepository->get($item->getSku())->getId();
```

Comment:
https://github.com/magento/magento2/blob/1ef4504e3c166ae0234dc3bb9ebe41fe26e90390/app/code/Magento/Catalog/Model/ResourceModel/Product.php#L617-L638

This would avoid a load in a loop?